### PR TITLE
Notifier API updates

### PR DESCRIPTION
**Pull request by michaelfig**
## Description

Improve typing (convey the type of a Notifier/Subscription via its SharableInternals) and tolerate remote AsyncIterables.

### Security Considerations

None.

### Documentation Considerations

Only changes as reflected in the public types.

### Testing Considerations

Passes existing tests.

### CODE REVIEW

File: packages/notifier/src/notifier.js
Comment (erights) [inline packages/notifier/src/notifier.js]: While it is true that this `.then` presents no hazard compared to the original `E.when`, I see two reasons three reasons to keep the old `E.when`:
   * Much programming is done by cargo culting. Someone reading this code and seeing the `.then` may not understand the subtle reasons why this one is safe, and be "inspired" to use a `.then` when it isn't safe. Since they already know `.then`, seeing one here rather than an `E.when` is a missed opportunity to help them unlearn.
   * Refactoring hazard: The delicate reason why this one is safe occurs on the previous line, with no obvious coupling. If we had linting or typing rules that could warn, ..., but we don't.
   * `E.when` accumulates the debugging traces needed for deep stacks. `.then` and `await` do not.

File: packages/notifier/src/notifier.js
Comment (erights) [inline packages/notifier/src/notifier.js:159]: `makeNotifierFromAsyncIterable` has a lot of the internal type decl comments that @dtribble  rightly point out we should minimize. I wrote most of them, but this PR adds yet another. I just tried the first obvious thing of removing each, and found that this new one is necessary, but the declarations on `optNextPromise`, `currentResponse`, and `baseNotifier` can be omitted without causing a failure to type check.

You are the master of jsdoc type checking and properjs, so I leave this to your judgement. No change needed as of this PR, especially since the new one as of this PR is necessary. But perhaps `makeNotifierFromAsyncIterable` will be an interesting case study of this issue going forward?

File: packages/notifier/src/notifier.js
Comment (erights) [inline packages/notifier/src/notifier.js]: This TODO will now appear in three places. It is appropriate in all three, or in none. I think it no longer makes sense, and so should be removed from all three rather than propagated.

File: packages/notifier/src/subscriber.js
Comment (erights) [inline packages/notifier/src/subscriber.js]: Why the underbar?

File: packages/notifier/src/types.js
Comment (erights) [inline packages/notifier/src/types.js]:
```suggestion
 * @typedef {ERef<BaseNotifier<T>>} NotifierInternals Will be shared between
 * machines, so it
```

and several more

File: packages/notifier/src/types.js
Comment (erights) [inline packages/notifier/src/types.js]: I purposely tried to make `SubscriptionInternals` opaque outside the package. Underbar is not a satisfying way to do that. Can we just give them normal names internally, but not declare them externally?

File: packages/notifier/src/types.js
Comment (erights) [inline packages/notifier/src/types.js]: What does "consistent" mean here? Comment please?

**Review by erights — CHANGES_REQUESTED** [review]

File: packages/notifier/src/notifier.js
Comment (michaelfig) [reply packages/notifier/src/notifier.js:159]: I think that annotating a declaration is perfectly fine, and not excessive.

The "type goo" that Dean refers to is usually more than one line of gymnastics to get existing code to conform to a type declaration.  I agree, and try to avoid that where possible.

File: packages/notifier/src/types.js
Comment (michaelfig) [reply packages/notifier/src/types.js]: I tried to minimise "type goo" by giving them external declarations because I wanted a way to convey the type parameter (`T`) of the notifier or subscription all the way through the sharable internals.

If you really want an opaque type, I can provide that, but it will cost some explicit casting in order to convey the parameter.

File: packages/notifier/src/notifier.js
Comment (erights) [inline packages/notifier/src/notifier.js]:
```suggestion
  const sharableInternalsP = 
    /** @type {ERef<SharableNotifierInternals<T>>} */ (unknownInternalsP);
```
This was well over 80 column limit, but I see `yarn lint-fix` doesn't fix it. Why not?

Comment (erights) [reply packages/notifier/src/notifier.js]: Several others too.

File: packages/notifier/src/types.js
Comment (erights) [inline packages/notifier/src/types.js:18]: Aha! Thanks for clarifying. I would not have guessed that.

File: packages/notifier/src/types.js
Comment (erights) [inline packages/notifier/src/types.js]: Wow, this is a lot more mechanism than I expected. Use your judgement about whether it is worth it. If not, revert as many of my suggested changes as needed to restore simplicity.

File: packages/notifier/test/iterable-testing-tools.js
Comment (erights) [inline packages/notifier/test/iterable-testing-tools.js:9]: Ah, I see. I thought it might be something from the assert package and was confused. I should have commented on it in my earlier review.

File: packages/notifier/src/notifier.js
Comment (erights) [reply packages/notifier/src/notifier.js:159]: @dtribble I'd appreciate your commentary, but in the meantime I'm resolving this conversation. @michaelfig since the type declarations at issue are not introduced by the PR anyway, do not await @dtribble 's commentary before merging.

**Review by erights — APPROVED** [review]
LGTM

File: packages/notifier/src/types.js
Comment (michaelfig) [reply packages/notifier/src/types.js]: > Wow, this is a lot more mechanism than I expected.

Yeah, TypeScript is not built to hide types from folks, it really wants to expose as many as possible.

For simplicity, I'll go back to transparent sharable internals.  But I will just use `head` and `tail` directly, not `_head` and `_tail` since that's a breaking change.

**Pull request merged**
